src/xrpld/app/misc/WasmHostFuncWrapper.cpp

```diff
@@ -38,70 +40,234 @@ setData(
+    if (it == m.end())
+    {
+        return Unexpected(HostFunctionError::INVALID_FIELD);
+    }
+    i++;
+    return *it->second;
+}
+
+template <>
+Expected<Slice, HostFunctionError>
+getData<Slice>(
+    InstanceWrapper const* rt,
+    wasm_val_vec_t const* params,
+    int32_t& i)
 {
+    auto const src = params->data[i].of.i32;
+    auto const ssz = params->data[i + 1].of.i32;
     if (src < 0 || ssz <= 0)
-        return Unexpected(HF_ERR_INVALID_PARAMS);
+        return Unexpected(HostFunctionError::INVALID_PARAMS);
 
     auto mem = rt ? rt->getMem() : wmem();
     if (!mem.s)
-        return Unexpected(HF_ERR_NO_MEM_EXPORTED);
+        return Unexpected(HostFunctionError::NO_MEM_EXPORTED);
 
     if (src + ssz > mem.s)
-        return Unexpected(HF_ERR_POINTER_OUT_OF_BOUNDS);
+        return Unexpected(HostFunctionError::POINTER_OUT_OF_BOUNDS);
 
-    Bytes data(mem.p + src, mem.p + src + ssz);
+    Slice data(mem.p + src, ssz);
+    i += 2;
     return data;
 }
 
-static Expected<AccountID, int32_t>
-getDataAccount(InstanceWrapper const* rt, int32_t ptr, int32_t sz)
+template <>
+Expected<uint256, HostFunctionError>
+getData<uint256>(
+    InstanceWrapper const* rt,
+    wasm_val_vec_t const* params,
+    int32_t& i)
 {
-    auto const r = getData(rt, ptr, sz);
+    auto const r = getData<Slice>(rt, params, i);
+    if (!r)
+    {
+        return Unexpected(r.error());
+    }
+
+    if (r->size() != uint256::bytes)
+    {
+        return Unexpected(HostFunctionError::INVALID_PARAMS);
+    }
+    return uint256::fromVoid(r->data());
+}
+
+template <>
+Expected<AccountID, HostFunctionError>
+getData<AccountID>(
+    InstanceWrapper const* rt,
+    wasm_val_vec_t const* params,
+    int32_t& i)
+{
+    auto const r = getData<Slice>(rt, params, i);
     if (!r || (r->size() != AccountID::bytes))
-        return Unexpected(HF_ERR_INVALID_PARAMS);
+        return Unexpected(HostFunctionError::INVALID_PARAMS);
 
     return AccountID::fromVoid(r->data());
 }
 
-static Expected<std::string, int32_t>
-getDataString(InstanceWrapper const* rt, int32_t src, int32_t ssz)
+template <>
+Expected<std::string_view, HostFunctionError>
+getData<std::string_view>(
+    InstanceWrapper const* rt,
+    wasm_val_vec_t const* params,
+    int32_t& i)
 {
+    auto const src = params->data[i].of.i32;
+    auto const ssz = params->data[i + 1].of.i32;
     if (src < 0 || ssz <= 0)
-        return Unexpected(HF_ERR_INVALID_PARAMS);
+        return Unexpected(HostFunctionError::INVALID_PARAMS);
 
     auto mem = rt ? rt->getMem() : wmem();
     if (!mem.s)
-        return Unexpected(HF_ERR_NO_MEM_EXPORTED);
+        return Unexpected(HostFunctionError::NO_MEM_EXPORTED);
 
     if (src + ssz > mem.s)
-        return Unexpected(HF_ERR_POINTER_OUT_OF_BOUNDS);
+        return Unexpected(HostFunctionError::POINTER_OUT_OF_BOUNDS);
 
     std::string data(mem.p + src, mem.p + src + ssz);
-    return data;
+    i += 2;
+    return std::string_view(data);
```

Comment:
This will be dangling string_view
getData<std::string_view>( InstanceWrapper const* rt, wasm_val_vec_t const* params,  int32_t& i)
{  auto const r = getData<Slice>(rt, params, i);
    if (!r)  return Unexpected(r.error());
    return std::string_view(reinterpret_cast<char const *>(r->data()), r->size());
    }